### ISSUE TITLE
RFC: Architectural Flaws of Cycle and Possible Solutions

### ISSUE THREAD

**Issue by edge** [cyclejs/core#268]
This issue is basically a compilation of stuff said in the [Gitter](https://gitter.im/cyclejs/core). Shoutout to @laszlokorte and @whitecolor for discussing these things with me. There are a lot of details, so feel free to point out and correct any mistakes.

**Before you read any of this, I want to say that idiomatic Cycle is currently neither truly pure nor truly fractal.**

### `isolate` is impure

`isolate` is necessary for ensuring two instances of a component are unique, but it also introduces a great amount of complexity to Cycle, and most drivers need to be aware of it. These circumstances create a lot of controversy over whether we should use `isolate` by default or if there is a better solution (e.g. #259).

Cycle's goal is to be pure and referentially transparent, but as long as we have `isolate`, it cannot be.

```js
isolate(Component) // impure
isolate(Component, scope) // pure <==> Component is pure
```

Since we can never be sure that a component is pure, it is safe to pessimistically assume that all components are impure.

The minimum requirements for a truly practical and pure `isolate` are threefold:
1. We must always pass an explicit scope to `isolate`
2. In order for this to scale, we must use a scope generator, from the top all the way down (humans are bound to create duplicates).

So far, the code would look like this:

(cc @theefer)

```js
function Component(sources) {
    const iso1 = sources.isolate;
    const [scope1, iso2] = iso1.next();
    const [scope2, iso3] = iso2.next();

    const { isolate: iso4 } = ChildA({ ...sources, isolate: iso3 });

    // ...
    return {
        // ...
        isolate: iso4
    };
}
```

3. This behavior is extremely redundant, and this purity criterion also requires us to pass it down as an extra sink on all components. The best way to practically do this is to use a pure functional language like Elm or Purescript, where we can then hide the `isolate` generator in a state monad.

### The Fractal Problem

André says that Cycle is a Fractal Architecture (http://staltz.com/unidirectional-user-interface-architectures.html). While this is true by interface, certain implementation details prevent it from being true in practice.

In Cycle's case, being fractal means that you can take any app's main function and pass it some drivers, or alternatively call a wrapped `run(main, drivers)` function, and expect it to behave normally.

But what if it uses `isolate`? We can't be sure that the child (possibly third party) component uses the same instance that we use. In the current source, `isolate` always counts up from zero. This means that two components using different `isolate` instances can clash. We've decidedly refused to use random number generators, since they is never truly safe. Using an incremental id is usually the safest bet, but only in a singular context, which we cannot assume.

One possible solution would be to somehow make sure that all components receive the same `isolate`, but that would only increase Cycle's overall complexity.

### Subjects to the rescue?

This idea came with some initial disapproval, but I've thought very much about it and I think it may be beneficial.

Instead of using `DOM.select`, we can use an event handler like in rx-recompose: https://github.com/acdlite/recompose/tree/master/src/packages/rx-recompose

(cc @vladap)

In the following code (which actually works), `DOM.handler` is equivalent to `createEventHandler` in rx-recompose. `createEventHandler` just returns a function that acts like `Subject`, but proxies calls to `onNext`. Using `$x$` is just a contrived convention, but the notation indicates that it is both an observer and an observable (i.e. subject).
```diff
export default ({ DOM }) => {
-	Button(DOM.select('.Home')).forEach(() => {
-		window.location.href = '/';
-	});
-	Button(DOM.select('.Github')).forEach(() => {
-		window.location.href = 'https://github.com/edge/cyc';
-	});
+	const $goToHome$ = DOM.handler();
+	const $goToGithub$ = DOM.handler();
+	$goToHome$.forEach(() => {
+		window.location.href = '/';
+	});
+	$goToGithub$.forEach(() => {
+		window.location.href = 'https://github.com/edge/cyc';
+	});
	return {
		DOM: $.just(
			div('.p2.measure', [
				h2('About'),
				h4([
					i('cyc'), ' is a Cycle.js boilerplate built with convenience and speed in mind.'
				]),
				br(),
-				button('.btn.Home', 'Home'), ' ',
-				button('.btn.Github', 'Github'),
+				button('.btn', { onclick: $goToGithub$ }, 'Home'), ' ',
+				button('.btn', { onclick: $goToGithub$ }, 'Github'),
			])
		)
	};
}
```

Although in the end you have two extra lines, this removes the need for `isolate`, drastically reducing complexity.

You might say that this is no longer pure, but `isolate` and Subjects are two sides of the same coin:
![http://i.imgur.com/gEJ2pxu.jpg](http://i.imgur.com/gEJ2pxu.jpg)

To the parent, the interface and behavior are the same. This follows two Cycle doctrines:
1. All read effects are from sources. All write effects are to sinks.
++ Subjects are implicit sources and sinks.
2. Parent sources go to Child sources, Child sinks go to Parent sinks.
++ As you can see in the diagram, no read-write arrow crosses a component context (lines in the diagram).

Not only is it more predictable, using subjects is about as pure as Javascript purity goes.

This also solves the fractal problem, because all Subjects are inherently unique.

What using this idiom implies for other drivers is an important topic for discussion, which I'll start off with this dialogue:
Q. How would this work for the HTTP driver?
A. `{ url, response$: $response$ }`
Q. But doesn't that violate the second doctrine? The subject will be passed into a driver, and a read event will bypass the parent sources.
A. Possibly, but if you look at it like metadata, which is more obvious as we are only passing a function containing a reference to the *actual* subject, it is no different from theoretical driver interop. The handler is an implicit source/sink, but we've conflated the two for simplicity. Although it may not seem to follow the doctrine, it is still fractal.

### But what about cycle-restart?

Using a subject for DOM events seems to exclude it from cycle-restart. In this case, we can add some code to the DOM driver to replay the subject's last event. This also solves https://github.com/Widdershin/cycle-restart/issues/41.

(cc @Widdershin)

If this concept becomes idiomatic, it could be standardized and applied to other drivers. For example, an HTTP.handler() could be implemented, mitigating the aforementioned context-crossing.

### The Fractal Problem (cont.)

This discussion does, however, bring up an important issue. Ideally, in an application implementing `cycle-restart`, every driver in an application should be `restartable`. This is the only way for an application to be truly restartable, as child applications with drivers would otherwise be ignored. In order for this to work, we need to explicitly pass down `rerun` to every component.

But what if a third party application uses its own instance of cycle-restart? This instance problem concerning cycle-restart, isolate, and other similar implementations *could* be partially solved by wrapping every application's `run(main, drivers)` as follows:
```js
export default ({ DOM, HTTP }) => {
    run(main, {
        DOM: DOM || restartable(makeDOMDriver('#root')),
        HTTP: HTTP || restartable(makeHTTPDriver())
    });
}
```
or possibly:
```js
export default (availableDrivers) => {
    run(main, {
        DOM: restartable(makeDOMDriver('#root')),
        HTTP: restartable(makeHTTPDriver())
    });
}
```
This means any available drivers would be reused, and would automatically be restartable. However, this would still require us to pass down `restart` and `isolate` in every component.

And what if we wanted hydratable drivers (not yet implemented)? Now *every* driver would also need to be wrapped in `hydratable`. This cannot scale.

Cycle seems to require a lot of passing-down-common-objects in order to be more correct as an architecture. While React solves this with `this.context` available to all impure components, we must solve it in a pure way to get closer to our goal of purity.

Monads are a viable solution; they are great for concisely and efficiently keeping things in a context, as well as reading and writing any necessary information. This implies that we should switch to a language that can facilitate this, like Elm or Purescript.

### Conclusion

1. Using Subjects instead of `isolate` would reduce a lot of complexity and make components much more predictable. It should live comfortably in both Javascript and compile-to-js environments.

2. The need for the switch to a functional language like Elm or Purescript to solve certain fundamental problems may be more imminent than we thought.

Gists:
1. https://gist.github.com/edge/5a5be8fed277bce2389e
2. https://gist.github.com/edge/39c1223f8f0e27a14e94

Further Reading:
1. https://gitter.im/cyclejs/core?at=56d7436c0bdb886502f6d263
2. https://gitter.im/cyclejs/core?at=56e5cd900055f8f35a82ccce
3. https://gitter.im/cyclejs/core?at=56e6002e89dd3cce10061e42
4. https://gitter.im/cyclejs/core?at=56e62ff13194fbd11096bb33

related: #259

**Comment by laszlokorte** [comment]
> (2) In order for this to scale, we must use a scope generator, from the top all the way down (humans are bound to create duplicates).

Isolation scopes do not have to be globally unique but only the sibling scopes inside one component must be unique because the names stack.
I do not think it is hard to come up with a handful of names. You have to to that all the times: variable names, function names, css classes...

> Since we can never be sure that a component is pure, it is safe to pessimistically assume that all components are impure.

We could just remove the second parameter for isolate. In my opinion it would not be a problem at all.
But sill you can not *guarantee* a function is pure just because it's javascript.
But taking that as a reason to assume everything is impure is flawed because than you could not use functions like `map` and `filter` at all because they assume you pass pure functions as well which as said you can not guarantee. It's all about conventions and trust.


> Q. How would this work for the HTTP driver?
A. { url, response$: $response$ }
Q. But doesn't that violate the second doctrine? The subject will be passed into a driver, and a read event will bypass the parent sources.

What you are re-inventing are kind of inverse callbacks. Instead of accepting input as parameter and returning a result like a function is meant to work you are returning "argument acceptors" that then need to be used to pass data into the functions.

As I wrote in twitter I can see that this get's rid of isolation but for me it sounds like turning:

```
function add(op1$, op2$) {
   return O.combineLatest(op1$, op2$, (a,b) => a + b);
}

add(O.just(5), O.just(7)).subscribe((r) => { /* r is 12 */ });
```

into:

```
function add() {
   const op1$ = Subject();
   const op2$ = Subject();
   
  return {
    op1$,
    op2$,
    result$: O.combineLatest(op1$, op2$, (a,b) => a + b);
  };
}

const adder = add();
adder.result$.subscribe((r) => { /* r is 12 */ });
O.just(5).subscribe(adder.op1$);
O.just(7).subscribe(adder.op2$);
```

In the first code uses a javascript function in a functional way: it transforms streams. In transforms input data into output data.

The second function builds up some *constant* observable graph. It does not use the concept of a function to it's full potential. It is more like a factory that constructs stream-graphs which then can be used.
It leads to the question why to use function arguments at all? You could also use subjects for props$ etc:

```
function MyCounter() {
   const props$ = Subject();
   const initialValue$ = Subject();
   const increment$ = Subject();
   const decrement$ = Subject();
   
  return {
    props$,
    initialValue$,
    increment$,
    decrement$
    value$: model(...);
  };
}
```

Of course it would work but it kind of misuses the language feature of a function and does not represent the data flow semantically correctly.

**Comment by edge** [comment]
> Isolation scopes do not have to be globally unique but only the sibling scopes inside one component must be unique because the names stack.

Right, but `.cycle-scope-cycle1.cycle-scope-cycle3` is the same as `.cycle-scope-cycle3.cycle-scope-cycle1` for the DOM driver. It's an unnecessary risk. And you can't be sure something like #243 won't happen, which is a real possibility.

> map and filter ... assume you pass pure functions

This is totally valid code:
```js
let multRandom$ = value$.map(v => Math.random() * v);
let dropRandom$ = value$.filter(() => Math.random() >= .5);
```

> What you are re-inventing are kind of inverse callbacks.

It's just copied over from rx-recompose.

> this get's rid of isolation but for me it sounds like turning: ...

They're different use cases. Your example shows how to collapse a run-of-the-mill use case, but you can't always do that for everything.

**Comment by Widdershin** [comment]
I think that there's value in discussing these ideas, but they overlook a key aspect of what I think is one of Cycle's biggest strengths, and the biggest reason to use it over Redux/Flux/Elm.

In Cycle, applications read from top to bottom in the following order:

* Given that changes to state are driven by these actions
* Change state
* Display state to the user

They are extremely easy to read once you're familiar, as you can always trace the source of where some data is flowing from just by reading the definitions.

With the ideas proposed in this issue, the order would be more like Redux/Flux/Elm.

* Given that some actions will happen, change state
* Display state to the user
* Handle user actions, trigger the above state changes

The difference there is that any part of your view can dispatch an action and change the application. You can't just trace back the flow of definitions to find all changes to state, you have to read the whole app.

In my opinion, `Subject` should not be used in application code. It makes it much harder to trace the flow of data in your app. They aren't declarative, and as a result usage isn't self describing.

Purity is a tradeoff. I think it's much more compelling to aim for the most useful feature of pure functions, writing deterministic applications.

`cycle-restart` demonstrates that a deterministic Cycle app where all changes to state are driven by external drivers allows you to treat the app as if it were a pure function.

The architecture problem I'm most excited for right now is how we declaratively express the relationships between nested dataflow objects, without using `Subject`. @staltz has said before that there's top/down and bottom up, but no middle. I'm not quite sure. I think there's room to investigate relational ideas and solve problems that way.

**Comment by staltz** [comment]
> I think that there's value in discussing these ideas, but they overlook a key aspect of what I think is one of Cycle's biggest strengths, and the biggest reason to use it over Redux/Flux/Elm.

Exactly my thoughts. We have been through these kind of issues before, and actually isolate came through one of these issues which criticize the current state Cycle.js  architecture. So while I really appreciate this and the suggestions and your time invested (plus diagrams), if what I say next sounds conservative it's because of the some concepts deeply embedded in the architecture.

> Purity is a tradeoff. I think it's much more compelling to aim for the most useful feature of pure functions, writing deterministic applications.

This is it, really.

When approaching these design issues, I'm always looking for an **accurate diagnosis** of the problem. So bear with me: what is really truly the problem of an impure `isolate()`?

- Is it simply the impurity? (Can explicit scope names solve it like Laszlo describe?)
- Is it naming collisions that may happen in fractals? (Can a better random naming scheme solve it? perhaps `cuid` or the global increment counter that works across execution context)

I don't believe impurity itself is the problem because you suggest using/creating Subjects which is inherently impure. So, is the second problem the real problem?

> @Widdershin  The architecture problem I'm most excited for right now is how we declaratively express the relationships between nested dataflow objects, without using Subject.

> @edge The best way to practically do this is to use a pure functional language like Elm or Purescript, where we can then hide the isolate generator in a state monad.

Yes I believe transitioning to a pure FP lang is the eventual goal. For the problem that @Widdershin mentioned, we currently solve it with Subjects because in JS the order of declarations matter, while in e.g. Haskell it doesn't.

- - -

Can I suggest we back we back off a bit from searching for solutions? Let's try to first pinpoint exactly what is the problem (and **why** is it an actual problem, what is it's root?), and let's define it in one simple sentence, and only after that look for solutions.

**Comment by laszlokorte** [comment]
> This is totally valid code:
> ```
let multRandom$ = value$.map(v => Math.random() * v);
let dropRandom$ = value$.filter(() => Math.random() >= .5);
```

Yeah but it's really bad code you would never want to write.

> It's just copied over from rx-recompose.

I know. and rx-recompose is based on react.js which.

> Right, but .cycle-scope-cycle1.cycle-scope-cycle3 is the same as .cycle-scope-cycle3.cycle-scope-cycle1 for the DOM driver. It's an unnecessary risk.

The stacking behavior of nested scopes is just an implementation detail. There may be slight issues at the moment (I am not even sure when the `.cycle-scope-cycle1.cycle-scope-cycle3` would occur).

> And you can't be sure something like #243 won't happen, which is a real possibility.

This can be fixed very easily by separating the namespace of generated and user provided scope names. (eg by prefixing)

----

As I have already written in the other discussion thread I agree with @staltz in this point: We can not solve a issue before the issue itself is well defined. Without someone giving a real good explanation of what the fundamental problem with `isolate()` is I see no point in discussing alternatives.

There are always different ways of doing something. Other frameworks like rx-recompose do it a different way. Well then why not just use a different framework? I mean there are not that much differences between cyclejs and rx-recompose. Both are lightweight wrappers around observables which allow for composition of UI components.

I think the choices staltz made when designing cycle are exactly what lead to the advantages cycle provides. If you do not like the way cycle works are it does not fit the demands of your project you are free to use another framework. And that's a good thing. I prefer 2 frameworks with different approaches over two frameworks with the same approach just because one copied the other for no good reason.

**Comment by staltz** [comment]
Yes to the final bit of Laszlo's message. Actually Yolk is a pretty good framework already. If you think that's what you want, go for it. It's good stuff. But Yolk and Cycle have different purposes. In the latter, it's an approachable and familiar way to writing components in a React-style, but using the power of Observables. In Cycle.js, the purpose is to provide a way of writing apps as a pure dataflow function from read effects to write effects, because this enables a one-to-one translation between your app's code and a signal graph.

**Comment by whitecolor** [comment]
I think using explicit `isolate` and really cyclic approach makes developer to **think more - and in my view that is a good thing** and using of `Subject` is seem to be a hack to me because there this no clear repsonsibility separation between READ/WRITE effects - and this is a conner stone in good design and architecture. 

People complain that cycle is not forcing to do some boilerplate and "care about things more". But if you don't want  to explicit, concise, logically verbose, you probably don't want to use cycle and declartive aproach in general. There is a lot of framewords that "hide" rough spots and do a lot of "magical" stuff (that actually often turns in to "magical" bugs).

There is a great article by @staltz http://staltz.com/its-easy-to-write-imperative.html. I think the more time and effort invested in thinking about apps logic, possible flaws and conner cases the less time in perspective will be required to support and modify this logic - this what functional programming and cycle in paticular it **just forces us to thing more**. From practice: even the people who don't know the cycle and actually are yet very bad at understanding functional approach and struggling with it feel by their gut that program becomes more reliable when they start to use this approach.

**Comment by edge** [comment]
> Is it naming collisions that may happen in fractals? (Can a better random naming scheme solve it? perhaps cuid or the global increment counter that works across execution context)

`cuid` did come across my mind as a potential solution, but I wonder about the performance impact of using many long classnames.

> [Subjects] overlook a key aspect of what I think is one of Cycle's biggest strengths, and the biggest reason to use it over Redux/Flux/Elm.

This was indeed one of my biggest inhibitions for using Subjects, but I believe the tradeoff between complexity and explicitness is still worth analyzing.

> I prefer 2 frameworks with different approaches over two frameworks with the same approach just because one copied the other for no good reason.

> In Cycle.js, the purpose is to provide a way of writing apps as a pure dataflow function from read effects to write effects, because this enables a one-to-one translation between your app's code and a signal graph.

In the end, Subjects are just a suggestion. What I really would like to see is explicit calls to `isolate` everywhere, where every `isolate` is passed a scope generated by on the previous call. We can then hide this behavior in a state monad, which is pretty easy to do in a functional language. I'm not sure how well Javascript would handle something like that.

**Comment by staltz** [comment]
> What I really would like to see is explicit calls to isolate everywhere, where every isolate is passed a scope generated by on the previous call.

You can already today start applying this technique, just giving that `next()` scope name to the 2nd arg of `isolate()`. You could try how that works and whether it's nice in developer ergonomics aspect.

But coming back to a previous topic: 

> Let's try to first pinpoint exactly what is the problem (and why is it an actual problem, what is it's root?), and let's define it in one simple sentence, and only after that look for solutions.

What is actually the problem of using the impure version `isolate`?

**Comment by Widdershin** [comment]
I should note that while working on cycle-restart I was originally very worried about `isolate` because using the implicit scope originally meant that the scope number would continue increasing every time the app restarts.

We added an `isolate.reset()` function to restart the count, it gets called when your app is restarted, and everything works just dandy.

It turned out that deterministic behavior was the most important part.

**Comment by staltz** [comment]
Good to keep in mind if we try to design alternative id generation methods.

**Comment by vladap** [comment]
@staltz You have criticized React for putting those data-id into dom, which I agree with and React as well as these are removed in React 0.15. Isolate seems to me to do pretty much the same thing. It puts cycle specific generated selectors into dom which leaks implementation detail into markup. What is the reasoning that it is fine now and React wasn't?

**Comment by vladap** [comment]
Cycle works directly with dom. But the current dom wasn't designed with components and isolation in mind.

1) What if we could do small changes to dom API so that it would support component isolation beatifully? Some kind of polyfill or we already use VDOM so some extension to it. And then start Cycle app from there instead of selecting directly from DOM CSS classess.

Whould each DOM element or component provide observable we could subscribe to? What would be your dream DOM API to run Cycle against?

2) Does incoming Web Components standard somehow help with this. i haven't read through it but maybe it already provides better API to isolate components.

**Comment by milankinen** [comment]
>Whould each DOM element or component provide observable we could subscribe to? What would be your dream DOM API to run Cycle against?

I've already told this few times in #259 but saying it once again: DOM events should be derived from the (V)DOM instead of generating the from "nothing". I mean something like [this](https://github.com/tsers-js/react#example).

**Comment by staltz** [comment]
@vladap good point indeed, and we intend to be less attached to classnames, see https://github.com/cyclejs/core/issues/226. About react-ids, they are usually all over the place (plus spans that wrapped every string), but cycle scope classes are only on the component boundaries.

**Comment by whitecolor** [comment]
This issue relates to possible isolation mechanics of HTTP driver
https://github.com/cyclejs/core/issues/272
